### ISSUE TITLE
Opening windows or menus increases CPU usage

### ISSUE THREAD

**Issue by Pentarctagon** [wesnoth/wesnoth#7615]
This was reported on Discord a few days ago where a player was having a significant performance issue when opening the recruit dialog (or other windows/menus), to the point where the reported FPS dropped to only 5. Additionally:
* They're using Windows, with a GeForce GTX 1660 Super GPU and a "3 GHz i7" CPU. I can reproduce the issue of increased CPU usage on Linux, however my CPU is apparently powerful enough that it doesn't affect my FPS.
* This doesn't happen on 1.16 and only started happening on 1.17 with 1.17.6. 1.17.6 is when hardware accelerated rendering was first enabled.
* GPU usage doesn't increase when this is happening, the fps just drops drastically.
* Disabling map and unit idle/standing animations fixes the issue.

**Comment by Pentarctagon** [comment]
@mesilliac if you're still around at all, do you have any idea what could be causing this?

**Comment by mesilliac** [comment]
Something is probably marking an area for redraw every frame. If it doesn't happen with animations disabled i would guess somehow the animation code is calling for a redraw, but animations are paused so it does it every frame without changing anything. That's just a wild guess however, as i never really touched the animation code.

**Comment by gfgtdf** [comment]
> but animations are paused

Actually they aren't (just tested wit the load game menu, standing animations still show), and i think this part is an improvement.

**Comment by gfgtdf** [comment]
I tested this, and the amount of cup increase does depend on the type of dialog:

No dialog: 1.5%
Chatlog : 6.5%
LoadGame: 30%
Statistics: 19%
StatusTable: 18.5%


So it seems in particular that the amount of cpu increase is not really proportinal to the window size (chatlog is bigger than Statistics) but to the complexity of the dialog (chatlog is a really  simple dialog.)

I can also confirm that this only happens when the dialog covers something animated (in particular water)

**Comment by gfgtdf** [comment]
I tested further with the load game dialog and i can for example also confirm that there is a small but consitent increase in cpu (26 to 30) usage whenever a savefile that shows a map is generated (as opposed to a start-of-scenario savefile that doesn't show a map).

**Comment by gfgtdf** [comment]
So if understood correctly there are 3 options for fix this issue:
1) Disable map animations while dialogs show
2) Disable map animations under the dialog only.
3) Optimize the gui2 drawing code further.

(1) Is probably the easiest to do, but (3) also has the chance to be quite easy if it turns out that there is some obvious bottleneck thats easy to fix.

I think the best thing to to is if someone could actually use a profiler to figure out whether there is an bottleneck that's easy to fix (i would know how to do it under windows with visualstudio, but not so sure about this linux computer)

**Comment by Pentarctagon** [comment]
By profiling, do you mean just sitting with the dialog open for a while?

**Comment by gfgtdf** [comment]
> By profiling, do you mean just sitting with the dialog open for a while?

Yes and (using a profiler program that tracks how much cpu is spend in the different parts of the code/functions.)

**Comment by Pentarctagon** [comment]
I did once use gperftools to profile config (https://github.com/wesnoth/wesnoth/issues/5368#issuecomment-858295242), so if that looks alright I can try using that to profile here.

**Comment by gfgtdf** [comment]
> I did once use gperftools to profile config ([#5368 (comment)](https://github.com/wesnoth/wesnoth/issues/5368#issuecomment-858295242)), so if that looks alright I can try using that to profile here.

Yes that would be nice i think, but it might be a more complicated since it's possible that its spends a lot of time in external libaries like pangocairo etc. (but its of course also possible that it doesn't), i'm also unsure how accurate -O0 information is actually for the user experience (which is i assume opinized builds)

**Comment by Pentarctagon** [comment]
What I recall is that using something other than `-O0` tends to result in gibberish in the profiling output. I can try both though and see what they look like.

**Comment by Pentarctagon** [comment]
Here's what it gave for sitting on Load Game for a while on Hamlets, built with -O3:
![loadgame](https://github.com/wesnoth/wesnoth/assets/3030250/2f48b75c-c39d-442c-921d-a0b13659d177)

**Comment by Pentarctagon** [comment]
And the statistics dialog:
![statistics](https://github.com/wesnoth/wesnoth/assets/3030250/f4d9770a-bd56-412e-bd61-637ad6be619f)

**Comment by Pentarctagon** [comment]
Steps for getting the profile images:
* installed libgoogle-perftools-dev, graphviz, google-perftools, libgoogle-perftools4
* built with `scons wesnoth --option-cache="" build=release --debug=time nls=false jobs=14 strict=false profiler=gperftools`
* ran the command `CPUPROFILE=test.prof ./wesnoth`
* and then generated the image with `google-pprof --gif ./wesnoth test.prof > ~/Downloads/loadgame.gif`

**Comment by gfgtdf** [comment]
So in your case the loadgame dialog spends most (50%) of its time in `image::render_minimap` and the statistics dialog spends most of time (40%) in `text_shape::draw()`, however when testing this, for me disabling the minimap (by selecting a save without a minimap), and caching the text renderer (and breaking some widgets in the process), only reduced cpu useage by about 15% (each) which is much less, so i suspect that there is also something else happening that is not not covered by your profiling, maybe in another internal thready by sdl.

Its also interesting to see that it spends most or at least a lot of the time in SDLs internals, which means that there is probably not that much we can do by optimizing "unrelated" parts of the code.


I think what we have to do is less invalidate_region calls, either try to omit hexes under a dialog or be a bit more smart/generous when uniting multiple  invalidated regions into one bigger region.

**Comment by gfgtdf** [comment]
> I think what we have to do is less invalidate_region calls, either try to omit hexes under a dialog or be a bit more smart/generous when uniting multiple invalidated regions into one bigger region.

I just tested by adding 
```
if(invalidated_.size() > 1) {
	draw_manager::invalidate_region(map_area());
}
```
here:
https://github.com/wesnoth/wesnoth/blob/0945c30efd2cafff9f36ca328bcd5ffc591e6be9/src/display.cpp#L2600
(which marks the whole map area as invalid instead of single tiles)
And it decreased the cpu usage of the loadgame dialog (during a game) from 30% to 13%.

**Comment by Pentarctagon** [comment]
Yeah, I see a similar decrease on my system.

I'm thinking of opening an issue over on the SDL2 github just asking why there's so much time spent in `__driDriverGetExtensions_d3d12` on Linux.

**Comment by gfgtdf** [comment]
hmm, usually i'd think github issues aren't the place for questions like that (imagine out bugtracker would be used for general wml questions :p), maybe they have a discord or mailinglist  where you could ask?

**Comment by gfgtdf** [comment]
But i'm actually also courious.

**Comment by mesilliac** [comment]
> I think what we have to do is less invalidate_region calls, either try to omit hexes under a dialog or be a bit more smart/generous when uniting multiple invalidated regions into one bigger region.

At the time i considered ignoring calls to redraw things that were completely occluded, but there were two problems:
1. if someone wanted to add a transparent or non-rectangular window they would want things underneath to draw
2. hexes don't fit nicely into a rectangular box, so there are almost always ones on the edge which are only partially occluded

(1) can mostly be ignored but (2) is a problem.

Re joining multiple invalidated regions into a bigger region, there is some code in `draw_manager` which tries to be clever about it, but it's difficult to know whether it's going to be cheaper to draw two smaller boxes or one much bigger box containing both of them. I don't remember what exact solution i chose for that but there is some code in there that can be tweaked.

>> I'd usually expect text rendering, but rendered text is cached so that shouldn't be expensive to redraw.
>
> Actually, are you sure about this? Text drawing seems to happen somewhere here 

It should be cached in the text rendering subsystem itself. If it gets a request for the same text in the same shape it just returns the cached result from last time.

**Comment by mesilliac** [comment]
Btw the modern solution to this is that the window or dialog renders itself once to a separate buffer, then when asked to redraw simply gives the result from that prerendered buffer. Then the compositor (the draw manager here) handles translucency effects etc.

Possibly the `invalidate_region` calls are assuming something like this, so are more aggressive about minimizing redrawn area than number of redraws.

**Comment by Pentarctagon** [comment]
@gfgtdf seems it's most likely an issue with the profiler.

**Comment by Pentarctagon** [comment]
The
```
if(invalidated_.size() > 1) {
	draw_manager::invalidate_region(map_area());
}
```
can probably be added, right?

For the minimap it seems like it's re-generating the texture constantly even though it rarely changes which means most of the work it does there is unnecessary, but given what you observed when disabling it I'm not sure if that'd be a meaningful change to make.

**Comment by gfgtdf** [comment]
> can probably be added, right?


Well it will probably increase cpu usage when no dialog is open. And its also not a very optimal solution. I think The least we should do is to only do that while a dialog is open, but the other solutions mentioned here would of course be better still.. But if we cannot find the time to do anythign better until 1.18 then maybe.

**Comment by mesilliac** [comment]
`draw_manager::invalidate_region` is indeed merging strictly according to minimum area.
https://github.com/wesnoth/wesnoth/blob/f5f76ade9a395a69cc21890ebc9ba0b9e442c38a/src/draw_manager.cpp#L90C1-L100C4

It could be changed to in stead always merge regions. Then you would have at most one redraw per frame. The tradeoff is that if you have some tiny thing change at the top left of the screen and some tiny thing change at the bottom right, it will redraw everything on the screen. This could be problematic for the game map for example, as it will redraw a lot of stuff that it doesn't need to all the time.

Better would probably be to add a "single-redraw" flag to `top_level_drawable`, and add a merge rule in `invalidate_region()` which adds that entire region if it's touched at all for any applicable TLDs. Then expensive things could specify they only want to be redrawn once at maxiumum.

**Comment by Pentarctagon** [comment]
I was able to get a bit of the SDL2 profiling data by building it locally, FWIW (load game dialog):
![perf-local](https://github.com/wesnoth/wesnoth/assets/3030250/3a16925f-e63f-493e-bb89-4e4697a824db)

**Comment by Pentarctagon** [comment]
And the Load Game dialog with the minimap disabled:
![perf-local](https://github.com/wesnoth/wesnoth/assets/3030250/1dd76b50-e5c6-492e-ad0b-1bb30517f5d6)

**Comment by Pentarctagon** [comment]
So given how much time gets spent calculating the size of text that never changes (I just open the Load Game dialog and then let it sit there for a few minutes), I'd assume it's not actually caching it.

**Comment by mesilliac** [comment]
Yeah looks like the text cache is for some reason not working effectively... and the minimap rerendering hurts.

The main game minimap caches its results, maybe this can too? Along with fixing the text caching that would probably solve this.

It might also not be all that hard to just ensure windows only rerender when they change. Reference for this is `display::render()` and `display::expose()`. The game display does cache its result, so `expose()` is just a blit from its private draw buffer. Draw code is mostly unchanged, it just uses a diffrent render target. Ideally everything would do the same, and in that case the text / minimap caching wouldn't matter so much.

I guess that would be the "improve gui2" option mentioned above.
Steps to implement: 
1. set up a private render texture for the drawable
2. render to it on `render()`, this can be skipped if nothing has changed
3. `expose()` is simply a blit from that buffer.

The game display uses two buffers but that's just to optimize map scrolling, so not necessary elsewhere.

Still, not sure how many gotchas there might be along the way.

**Comment by gfgtdf** [comment]
> Better could be to add a "single-redraw" flag to top_level_drawable

This would mean that also single buttons etc, always redraw the whole dialog tho.

**Comment by gfgtdf** [comment]
> I was able to get a bit of the SDL2 profiling data by building it locally

It still doesn't tell you why it ends up calling the d3d functions tho :p, That probably somewhere into the olg implementation, probably in the driver software of your graphic card, we don't even know whether that part is open source.

> The main game minimap caches its results, maybe this can too?

At this point i want to remind that the ingame minimap also has problems #1605 and also the the minimap widget has other problems: #7087

> Steps to implement:

I think at least for the upcoming 1.18 release we should probably go with the other option, but this is sureley somethign that makes sense to do later (if someone finds the time for it of course)

**Comment by gfgtdf** [comment]
> I tested this, and the amount of cup increase does depend on the type of dialog:
> 
> No dialog: 1.5% Chatlog : 6.5% LoadGame: 30% Statistics: 19% StatusTable: 18.5%
> 
> So it seems in particular that the amount of cpu increase is not really proportinal to the window size (chatlog is bigger than Statistics) but to the complexity of the dialog (chatlog is a really simple dialog.)
> 
> I can also confirm that this only happens when the dialog covers something animated (in particular water)

Regarding Loadgame dialog performance i should notice that i usually play with altternative minimap terrain Coloring enabled (the most right button under the ingame minimap), with default minimap terrain coloring the cpu usage during loadgame dialog  increases from 30% to 40%.

**Comment by mesilliac** [comment]
I had a quick look to see how hard it would be to implement the offscreen rendering, and it was basically straightforward. I just copied the display code to render to a buffer for all gui2 windows. I think that should solve this problem and any others like it.

I don't have time to test it much but it seemed to work and completely resolved this CPU issue for me.

**Comment by mesilliac** [comment]
Should be fixed by #7982
